File: frontend/src/lib/billing.ts
import { api } from '$lib/api/client';
import type { components } from '$lib/api/schema';

export type BillingStatus = components['schemas']['BillingStatusResponse'];
export type BillingSummary = components['schemas']['BillingSummary'];

// Fetch the active org's billing status. Returns null when unauthenticated or
// billing is unavailable.
export async function fetchBilling(): Promise<BillingStatus | null> {
	const { data, error } = await api.GET('/billing');
	if (error || !data) return null;
	return data as BillingStatus;
}

// Start a Stripe Checkout session for the Pro plan. Returns the URL to
// redirect the browser to, or an error message.
export async function createCheckout(): Promise<{ url?: string; error?: string }> {
	const { data, error } = await api.POST('/billing/checkout');
	if (error) return { error: error.message ?? 'failed to start checkout' };
	if (!data?.url) return { error: 'failed to start checkout' };
	return { url: data.url };
}

// Open the Stripe Customer Portal (card updates, cancellation, invoices).
export async function createPortal(): Promise<{ url?: string; error?: string }> {
	const { data, error } = await api.POST('/billing/portal');
	if (error) return { error: error.message ?? 'failed to open the billing portal' };
	if (!data?.url) return { error: 'failed to open the billing portal' };
	return { url: data.url };
}

// Apply for the free open-source tier. Returns the refreshed billing status,
// or an error message.
export async function submitOssApplication(
	sponsorUrl: string,
	note?: string
): Promise<{ status?: BillingStatus; error?: string }> {
	const { data, error } = await api.POST('/billing/oss-application', {
		body: { sponsorUrl, ...(note ? { note } : {}) }
	});
	if (error) return { error: error.message ?? 'failed to submit the application' };
	return { status: data as BillingStatus };
}

// Days (rounded up, floor 0) until the trial ends.
export function trialDaysLeft(trialEndsAt: string): number {
	const ms = new Date(trialEndsAt).getTime() - Date.now();
	return Math.max(0, Math.ceil(ms / (24 * 60 * 60 * 1000)));
}

// "$9.99" from cents.
export function formatPrice(cents: number): string {
	return `$${(cents / 100).toFixed(2)}`;
}

// Human label for a plan key.
export function planLabel(plan: string): string {
	switch (plan) {
		case 'beta':
			return 'Beta';
		case 'pro':
			return 'Pro';
		case 'oss_free':
			return 'Open Source';
		case 'enterprise':
			return 'Enterprise';
		default:
			return 'Trial';
	}
}

File: backend/internal/httpapi/billing.go
package httpapi

import (
	"context"
	"errors"
	"log/slog"
	"strings"

	"xolo/backend/internal/api"
	"xolo/backend/internal/auth"
	"xolo/backend/internal/billing"
)

// billingAdminOnlyMsg is the 403 body for billing actions gated to admins.
const billingAdminOnlyMsg = "only admins can manage billing"

// GetBilling implements `getBilling`: GET /billing.
// Returns the active organization's billing status, lazily starting its
// 21-day trial on first touch, and trues the Stripe seat quantity up while a
// subscription is live (the billing page is one of the seat-sync paths).
func (h Handler) GetBilling(ctx context.Context) (api.GetBillingRes, error) {
	user := auth.UserFromContext(ctx)
	if user == nil {
		return &api.GetBillingUnauthorized{Message: "unauthorized"}, nil
	}
	if user.OrgID == "" {
		return &api.GetBillingBadRequest{Message: "no active organization"}, nil
	}
	h.billing.ReconcileSeats(ctx, user.OrgID)
	status, err := h.billing.StatusForOrg(ctx, user.OrgID)
	if err != nil {
		slog.ErrorContext(ctx, "httpapi: billing status failed", "org_id", user.OrgID, "error", err)
		return &api.GetBillingBadRequest{Message: "billing is not configured"}, nil
	}
	return toBillingStatusResponse(status), nil
}

// CreateBillingCheckout implements `createBillingCheckout`: POST /billing/checkout.
// Starts a subscription-mode Stripe Checkout session for the Pro plan.
func (h Handler) CreateBillingCheckout(ctx context.Context) (api.CreateBillingCheckoutRes, error) {
	user := auth.UserFromContext(ctx)
	if user == nil {
		return &api.CreateBillingCheckoutUnauthorized{Message: "unauthorized"}, nil
	}
	if user.OrgID == "" {
		return &api.CreateBillingCheckoutBadRequest{Message: "no active organization"}, nil
	}
	if user.Role != auth.RoleAdmin {
		return &api.CreateBillingCheckoutForbidden{Message: billingAdminOnlyMsg}, nil
	}
	url, err := h.billing.CreateCheckout(ctx, user.OrgID, user.Email)
	switch {
	case errors.Is(err, billing.ErrDisabled):
		return &api.CreateBillingCheckoutBadRequest{Message: "billing is disabled while NotifBuddy is in beta"}, nil
	case errors.Is(err, billing.ErrAlreadySubscribed):
		return &api.CreateBillingCheckoutConflict{Message: "already subscribed"}, nil
	case errors.Is(err, billing.ErrNotConfigured):
		return &api.CreateBillingCheckoutBadRequest{Message: "billing is not configured"}, nil
	case err != nil:
		slog.ErrorContext(ctx, "httpapi: create checkout failed", "org_id", user.OrgID, "error", err)
		return &api.CreateBillingCheckoutBadRequest{Message: "failed to start checkout"}, nil
	}
	return &api.BillingRedirectResponse{URL: url}, nil
}

// CreateBillingPortal implements `createBillingPortal`: POST /billing/portal.
// Opens the Stripe Customer Portal for the org's customer.
func (h Handler) CreateBillingPortal(ctx context.Context) (api.CreateBillingPortalRes, error) {
	user := auth.UserFromContext(ctx)
	if user == nil {
		return &api.CreateBillingPortalUnauthorized{Message: "unauthorized"}, nil
	}
	if user.OrgID == "" {
		return &api.CreateBillingPortalBadRequest{Message: "no active organization"}, nil
	}
	if user.Role != auth.RoleAdmin {
		return &api.CreateBillingPortalForbidden{Message: billingAdminOnlyMsg}, nil
	}
	url, err := h.billing.CreatePortal(ctx, user.OrgID)
	switch {
	case errors.Is(err, billing.ErrDisabled):
		return &api.CreateBillingPortalBadRequest{Message: "billing is disabled while NotifBuddy is in beta"}, nil
	case errors.Is(err, billing.ErrNoCustomer):
		return &api.CreateBillingPortalBadRequest{Message: "no billing account yet — subscribe first"}, nil
	case errors.Is(err, billing.ErrNotConfigured):
		return &api.CreateBillingPortalBadRequest{Message: "billing is not configured"}, nil
	case err != nil:
		slog.ErrorContext(ctx, "httpapi: create portal failed", "org_id", user.OrgID, "error", err)
		return &api.CreateBillingPortalBadRequest{Message: "failed to open the billing portal"}, nil
	}
	return &api.BillingRedirectResponse{URL: url}, nil
}

// SubmitOssApplication implements `submitOssApplication`: POST /billing/oss-application.
// Records a free open-source tier application (manually reviewed).
func (h Handler) SubmitOssApplication(ctx context.Context, req *api.OssApplicationRequest) (api.SubmitOssApplicationRes, error) {
	user := auth.UserFromContext(ctx)
	if user == nil {
		return &api.SubmitOssApplicationUnauthorized{Message: "unauthorized"}, nil
	}
	if user.OrgID == "" {
		return &api.SubmitOssApplicationBadRequest{Message: "no active organization"}, nil
	}
	if user.Role != auth.RoleAdmin {
		return &api.SubmitOssApplicationForbidden{Message: billingAdminOnlyMsg}, nil
	}
	sponsorURL := strings.TrimSpace(req.SponsorUrl)
	if !strings.HasPrefix(sponsorURL, "https://") && !strings.HasPrefix(sponsorURL, "http://") {
		return &api.SubmitOssApplicationBadRequest{Message: "sponsorUrl must be a http(s) URL"}, nil
	}
	note := ""
	if n, ok := req.Note.Get(); ok {
		note = strings.TrimSpace(n)
	}
	status, err := h.billing.SubmitOSSApplication(ctx, user.OrgID, sponsorURL, note)
	switch {
	case errors.Is(err, billing.ErrDisabled):
		return &api.SubmitOssApplicationBadRequest{Message: "billing is disabled while NotifBuddy is in beta"}, nil
	case errors.Is(err, billing.ErrAlreadySubscribed):
		return &api.SubmitOssApplicationConflict{Message: "already approved or subscribed"}, nil
	case err != nil:
		slog.ErrorContext(ctx, "httpapi: oss application failed", "org_id", user.OrgID, "error", err)
		return &api.SubmitOssApplicationBadRequest{Message: "failed to record the application"}, nil
	}
	return toBillingStatusResponse(status), nil
}

// toBillingStatusResponse maps a derived billing status to the generated type.
func toBillingStatusResponse(st billing.Status) *api.BillingStatusResponse {
	resp := &api.BillingStatusResponse{
		Plan:              api.BillingStatusResponsePlan(st.Plan),
		Locked:            st.Locked,
		TrialEndsAt:       st.TrialEndsAt,
		Subscribed:        st.Subscribed,
		PriceCentsPerSeat: billing.PriceCentsPerSeat,
	}
	if st.StripeStatus != "" {
		resp.StripeStatus = api.NewOptString(st.StripeStatus)
	}
	if st.Seats > 0 {
		resp.Seats = api.NewOptInt(st.Seats)
	}
	if st.OSSApplicationStatus != "" {
		resp.OssApplicationStatus = api.NewOptBillingStatusResponseOssApplicationStatus(
			api.BillingStatusResponseOssApplicationStatus(st.OSSApplicationStatus))
	}
	return resp
}
